### ISSUE TITLE
MaterialApp should support unconstrained environments

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#137875]
If an Overlay widget finds itself in an unconstrained environment (e.g. because the View is allowed to pick a dynamic size, see https://github.com/flutter/flutter/issues/134501), it throws. This means, MaterialApp (which introduces an Overlay as part of the Navigator) cannot be placed in an unconstrained view. We somehow need to allow that. I suspect, it means limiting the available options to position things inside the overlay...

**Comment by goderbauer** [comment]
FYI @ditman

**Comment by goderbauer** [comment]
This is actually more complicated. It's not just the overlay that blows up in unconstraint environments. Things like Scaffold (and anything else that uses a CustomMultiChildLayout amongst other things) also cannot live in unconstraint environments. So, just fixing Overlay is not gonna be enough in the long term.

cc @yjbanov @ditman Do you have an idea of what kind of UIs need to work in unconstrained environments so we can prioritize those? I am afraid that just fixing the Overlay doesn't get us all that far...

**Comment by ditman** [comment]
> @ditman Do you have an idea of what kind of UIs need to work in unconstrained environments so we can prioritize those?

Now that the constraints for the root view can be set from the embedder, potentially any app can end up in an unconstrained view.

One simple example that I couldn't get to work, is adding a [SelectionArea](https://api.flutter.dev/flutter/material/SelectionArea-class.html) to your "yellow box with text inside" reflow sample, because SelectionArea ends up needing an Overlay (that's the first bug I've noticed regarding this, there's maybe others)

**Comment by goderbauer** [comment]
I've been looking into getting the counter app (the app you get when you do `flutter create`) to work in an unconstrained view. It uses some of the problematic widgets like Overlay and Scaffold.

> (PS: Do we need to re-think how views are able to resize from the framework, or would we end up in a similar situation anyways?)

I believe the issue is inherent to the problem we are trying to solve. We want to size the view based on its content. However, currently some widgets (Overlay, Scaffold, to name some) are unable to size themselves based on their content. They rely on the parent telling them what size they should be (but of course if the size should be based on content the parent doesn't know what size to give the child).

Some of this is due to history. There was never a reason for some of these widgets to size themselves based on content because they could just rely on the fact that they will be used in a fixed size view that needs to be filled out. For some of these widgets this problem is straight-forward to fix:

If the Overlay finds itself in an unconstraint environment, it could pick a child that determines its size (e.g. the topmost child that wants to fill the entire view is a good candidate). I've implemented this in https://github.com/flutter/flutter/pull/139513.

For other widgets, this is a lot harder to fix. The Scaffold is a good example. For simplicity, lets assume our Scaffold just has an app bar and a body. They are obviously expected to have the same width, but in order to figure out what width that should be in an unconstrained environment we would need to lay out both, measure their individual widths, pick the larger one and lay them both out again at that desired overall width. We are in intrinsics sizing territory here, which comes with performance penalties due the additional layout passes required. Other widgets based on CustomMultiChildLayout and CustomSingleChildLayout face the same problem.

If we want to make this work, we likely will not get around the fact that this may require intrinsics calculations. One thought I had to make this work: Instead of fixing up each individual widget (as illustrated above for the Scaffold), can we just wrap the entire widget tree in IntrinsicsWidth/IntrinsicsHeight widgets? In theory, this should work. In practice we face the problem that widgets based on CustomMultiChildLayout don't implement the intrinsics protocol properly (see [TODO](https://github.com/flutter/flutter/blob/bbe980cfe88a3b115cdb763aaf91e80eb8df0544/packages/flutter/lib/src/rendering/custom_layout.dart#L355-L357)), likely again because they never had to work in these kind of environments before. So, as before we'd be back to fixing up the intrinsics of the individual widgets.

Furthermore, even if we fix up intrinsics for these widgets we still have the fundamental limitation that intrinsics do not work for LayoutBuilder-type widgets (and until https://github.com/flutter/flutter/pull/138369 is fixed it also doesn't work for anything that requires baseline calculations). So, those types of widgets couldn't be used inside an unconstrained view (unless they have a parent widget that gives them constrained constraints).

Long story short: I don't think we will ever be able to allow any kind of widget inside an unconstrained view. Even getting a reasonable generic subset to work is gonna require a heavy lift of fixing up assumptions/bugs in our existing widgets. For a specific UI where the number of widgets that need to be fixed is clear, this may be possible, but likely in the short term not for generic "unknown" UIs. 🫤
